### PR TITLE
Increase limit on allowed values of intra-period

### LINKED ISSUE #1492: Allow GOP sizes larger than 256
Currently, the maximum keyint is 256. This is probably fine for footage that's 30 fps or below as it allows a keyframe interval of up to 10-ish seconds, but for 60 fps and up it starts to run a bit short.

Also, it's a bit counterintuitive that the GOP size is always one larger than the value input with `--keyint`. I.e. `--keyint 60` results in a GOP size of 61.

### PR DESCRIPTION
**Pull request by ccccheung**
# Description
Before, we had intra-period was constrained to 255 which limited the maximum gop size to 256.
I have increased the maximum gop size to the limit of int32_t.

Tests done:
BD-rate comparison with 6 60 frame 8 bit and 6 60 frame 10 bit clips with a mix of resolutions with intra period 119
at M1, M4, M8
   - Found no performance impact (quality or speed)

Same comparison as above except with intra-period 50,000
   - Found bitstreams produced with the same, and no performance impact

Encoded a 301 frame clip and encoded with the following intra periods:
   - -1, 100, 200, 280,290,301, 3000,50000
the 100, 200, 280, 290 encode generated different bitstreams from one another, the -1, 301, 3000, and 50000 encode generated the same bitstream as each other as expected.
I also verified that a keyframe was inserted at frame 281 for the 280 encode.

BD-rate comparison using no-lp (M0, M4,M8)  and 2 pass CRF (M0 and M3) 
     -There is no impact on performance when using intra-period 50,000

# Issue
Fixes #1492 

# Author(s)
Colton Cheung @ccccheung 

# Performance impact

- [ ] quality
- [ ] memory
- [ ] speed
- [ ] 8 bit
- [ ] 10 bit
- [x] N/A

# Test set
- [ ] obj-1-fast can be found [here](https://media.xiph.org/video/derf/objective-1-fast.tar.gz)
- [ ] other
- [x] N/A

# Merge method
- [x] Allow the maintainer to squash and merge when PR is ready to create a 1-commit to the master branch. The maintainer will be able to fix typos / combine commit messages to create a more readable 1-commit message or use whatever is stated in the 'Description' section
- [ ] I will clean up my commits and the maintainer shall use 'rebase and merge' to the master branch

### CODE REVIEW

File: Docs/svt-av1_encoder_user_guide.md [hunk 1]
```diff
@@ -190,7 +190,7 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
 #### Keyframe Placement Options
 | **Configuration file parameter** | **Command line** | **Range** | **Default** | **Description** |
 | --- | --- | --- | --- | --- |
-| **IntraPeriod** | --keyint | [-2 - 255] | -2 | Intra period interval(frames) -2: default intra period , -1: No intra update or [0-255] |
+| **IntraPeriod** | --keyint | [-2 - 2^32-2] | -2 | Intra period interval(frames) -2: default intra period , -1: No intra update or [0-2147483646] |
```
Comment (hassount) [inline Docs/svt-av1_encoder_user_guide.md:193]: Can we change the 2147483646 to be a bit more abstract, 2^32-2 or maybe >= 0?

File: Config/Sample.cfg [hunk 2]
```diff
@@ -53,7 +53,7 @@ AdaptiveQuantization            : 0                         # Set adaptive QP le
 VBVBufSize                      : 0                         # VBV buffer size
 
 #=============================== Keyframe Placement Options ===============================
-IntraPeriod                     : -2                        # Intra period interval(frames) (-2: default intra period, -1: No intra update or [0-255])
+IntraPeriod                     : -2                        # Intra period interval(frames) (-2: default intra period, -1: No intra update or [0-2147483646])
```
Comment (hassount) [inline Config/Sample.cfg:56]: Can we change the 2147483646 to be a bit more abstract, 2^32-2 or maybe >= 0

File: Source/App/EncApp/EbAppConfig.c [hunk 3]
```diff
@@ -856,11 +856,11 @@ ConfigEntry config_entry_intra_refresh[] = {
     // File I/O
     //{SINGLE_INPUT,
     // INTRA_PERIOD_TOKEN,
-    //"Intra period interval(frames) (-2: No intra update, -1: default intra period or [0-255])",
+    //"Intra period interval(frames) (-2: No intra update, -1: default intra period or [0-2147483646])",
```
Comment (hassount) [inline Source/App/EncApp/EbAppConfig.c:859]: more changes needed here maybe ```2^32-2``` ?

File: Source/App/EncApp/EbAppConfig.c [hunk 4]
```diff
@@ -856,11 +856,11 @@ ConfigEntry config_entry_intra_refresh[] = {
...
     // set_cfg_intra_period},
     {SINGLE_INPUT,
      KEYINT_TOKEN,
-     "Intra period interval(frames) (-2: default intra period , -1: No intra update or [0-255])",
+     "Intra period interval(frames) (-2: default intra period , -1: No intra update or [0-2147483646])",
```
Comment (hassount) [inline Source/App/EncApp/EbAppConfig.c:863]: ditto

File: Source/App/EncApp/EbAppConfig.c [hunk 5]
```diff
@@ -1318,7 +1318,7 @@ ConfigEntry config_entry[] = {
     {SINGLE_INPUT, INPUT_DEPTH_TOKEN, "Bit depth for codec(8 or 10)", set_encoder_bit_depth},
     {SINGLE_INPUT,
      KEYINT_TOKEN,
-     "Intra period interval(frames) (-2: No intra update, -1: default intra period or [0-255])",
+     "Intra period interval(frames) (-2: No intra update, -1: default intra period or [0-2147483646])",
```
Comment (hassount) [inline Source/App/EncApp/EbAppConfig.c:1321]: ditto

File: Source/Lib/Encoder/Globals/EbEncHandle.c [hunk 6]
```diff
@@ -2381,8 +2381,8 @@ static EbErrorType verify_settings(
         SVT_LOG("Error instance %u: Hierarchical Levels supported [3-5]\n", channel_number + 1);
         return_error = EB_ErrorBadParameter;
     }
-    if (config->intra_period_length < -2 || config->intra_period_length > 255) {
-        SVT_LOG("Error Instance %u: The intra period must be [-2 - 255] \n", channel_number + 1);
+    if (config->intra_period_length < -2 || config->intra_period_length > 2147483646) {
```
Comment (hassount) [inline Source/Lib/Encoder/Globals/EbEncHandle.c:2384]: ditto

**Comment by silverbacknet** [comment]
The magic number is referenced enough that it should probably be a #define or just remove it from the help text, it's so huge. 0 as intra-only might be worth calling out.

What's the deal with lines 863 and 1321 having reversed -1/-2 semantics?

File: Source/App/EncApp/EbAppConfig.c (hunk 3 above)
Comment (hassount) [reply Source/App/EncApp/EbAppConfig.c:859]: actually this part of the code is commented out and should probably be removed

**Comment by hassount** [comment]
> What's the deal with lines 863 and 1321 having reversed -1/-2 semantics?

https://github.com/AOMediaCodec/SVT-AV1/pull/1497/files#r493030932

**Comment by silverbacknet** [comment]
> 
> 
> > What's the deal with lines 863 and 1321 having reversed -1/-2 semantics?
> 
> https://github.com/AOMediaCodec/SVT-AV1/pull/1497/files#r493030932

OK, so -1 is definitely no-intra, but SvtAv1EncApp -help currently prints `--keyint                        Intra period interval(frames) (-2: No intra update, -1: default intra period or [0-255])`

Let's deal with this and I'll send a pull request to fix that.

**Comment by hassount** [comment]
> > > What's the deal with lines 863 and 1321 having reversed -1/-2 semantics?
> > 
> > 
> > https://github.com/AOMediaCodec/SVT-AV1/pull/1497/files#r493030932
> 
> OK, so -1 is definitely no-intra, but SvtAv1EncApp -help currently prints `--keyint Intra period interval(frames) (-2: No intra update, -1: default intra period or [0-255])`
> 
> Let's deal with this and I'll send a pull request to fix that.

I thought this is what's currently in the help menu, I have fixed this recently:

```
"Intra period interval(frames) (-2: default intra period , -1: No intra update or [0-255])"
```

**Comment by ccccheung** [comment]
> > > > What's the deal with lines 863 and 1321 having reversed -1/-2 semantics?
> > > 
> > > 
> > > https://github.com/AOMediaCodec/SVT-AV1/pull/1497/files#r493030932
> > 
> > 
> > OK, so -1 is definitely no-intra, but SvtAv1EncApp -help currently prints `--keyint Intra period interval(frames) (-2: No intra update, -1: default intra period or [0-255])`
> > Let's deal with this and I'll send a pull request to fix that.
> 
> I thought this is what's currently in the help menu, I have fixed this recently:
> 
> ```
> "Intra period interval(frames) (-2: default intra period , -1: No intra update or [0-255])"
> ```

It's correct for the single dash token, incorrect for the double dash token. I can fix this will I'm at it.

File: Config/Sample.cfg [hunk 7]
```diff
@@ -53,7 +53,7 @@ AdaptiveQuantization            : 0                         # Set adaptive QP le
 VBVBufSize                      : 0                         # VBV buffer size
 
 #=============================== Keyframe Placement Options ===============================
-IntraPeriod                     : -2                        # Intra period interval(frames) (-2: default intra period, -1: No intra update or [0-255])
+IntraPeriod                     : -2                        # Intra period interval(frames) (-2: default intra period, -1: No intra update or an integer >=0)
```
Comment (hassount) [inline Config/Sample.cfg:56]: can you please add the limit of 255 for rc 1/2 in the documentation?

File: Source/Lib/Encoder/Globals/EbEncHandle.c [hunk 8]
```diff
@@ -2381,8 +2381,13 @@ static EbErrorType verify_settings(
         SVT_LOG("Error instance %u: Hierarchical Levels supported [3-5]\n", channel_number + 1);
         return_error = EB_ErrorBadParameter;
     }
-    if (config->intra_period_length < -2 || config->intra_period_length > 255) {
-        SVT_LOG("Error Instance %u: The intra period must be [-2 - 255] \n", channel_number + 1);
+    if ((config->intra_period_length < -2 || config->intra_period_length > 2147483646) && config->rate_control_mode == 0) {
```
Comment (hassount) [inline Source/Lib/Encoder/Globals/EbEncHandle.c:2384]: maybe 1<<32 -2 would be better than 2147483646?

**Comment by paulsgh** [comment]
Tested a mix of clips comprised of different resolutions and bit depths.
This PR is lossless.

| | Speed Deviation  | BD-Rate   Deviation | Memory Deviation
-- | -- | -- | --
M0 - 8bit - 2pass| 1.07% | 0.00% | 0.00%
M0 - 10bit - 2pass| 0.77% | 0.00% | -0.01%
M0 - 8bit - 1pass| 1.05% | 0.00% | -0.01%
M4 - 8bit - 1pass| 1.01% | 0.00% | 0.02%
M8 - 8bit - 1pass| 0.91% | 0.00% | -0.01%
M0 - 10bit - 1pass| 0.81% | 0.00% | 0.00%
M4 - 10bit - 1pass| 1.04% | 0.00% | -0.01%
M8 - 10bit - 1pass| 0.24% | 0.00% | -0.03%

**Pull request merged**